src/auth/flows.rs

```diff
@@ -1438,9 +1561,385 @@ pub async fn remove_2fa(
+                .await?
+                .ok_or_else(|| AuthenticationError::InvalidCredentials)?;
+
+        if !scopes.contains(Scopes::USER_AUTH_WRITE) {
+            return Err(ApiError::Authentication(
+                AuthenticationError::InvalidCredentials,
+            ));
+        }
+
+        if let Some(pass) = user.password.as_ref() {
+            let old_password = change_password.old_password.as_ref().ok_or_else(|| {
+                ApiError::CustomAuthentication(
+                    "You must specify the old password to change your password!".to_string(),
+                )
+            })?;
+
+            let hasher = Argon2::default();
+            hasher.verify_password(old_password.as_bytes(), &PasswordHash::new(pass)?)?;
+        }
+
+        user
+    };
+
+    let mut transaction = pool.begin().await?;
+
+    let update_password = if let Some(new_password) = &change_password.new_password {
+        let score = zxcvbn::zxcvbn(
+            new_password,
+            &[
+                &user.username,
+                &user.email.clone().unwrap_or_default(),
+                &user.name.unwrap_or_default(),
+            ],
+        )?;
+
+        if score.score() < 3 {
+            return Err(ApiError::InvalidInput(
+                if let Some(feedback) = score.feedback().clone().and_then(|x| x.warning()) {
+                    format!("Password too weak: {}", feedback)
+                } else {
+                    "Specified password is too weak! Please improve its strength.".to_string()
+                },
+            ));
+        }
+
+        let hasher = Argon2::default();
+        let salt = SaltString::generate(&mut ChaCha20Rng::from_entropy());
+        let password_hash = hasher
+            .hash_password(new_password.as_bytes(), &salt)?
+            .to_string();
+
+        Some(password_hash)
+    } else {
+        if !(user.github_id.is_some()
+            || user.gitlab_id.is_some()
+            || user.microsoft_id.is_some()
+            || user.google_id.is_some()
+            || user.steam_id.is_some()
+            || user.discord_id.is_some())
+        {
+            return Err(ApiError::InvalidInput(
+                "You must have another authentication method added to remove password authentication!".to_string(),
+            ));
+        }
+
+        None
+    };
+
+    sqlx::query!(
+        "
+        UPDATE users
+        SET password = $1
+        WHERE (id = $2)
+        ",
+        update_password,
+        user.id as crate::database::models::ids::UserId,
+    )
+    .execute(&mut *transaction)
+    .await?;
+
+    if let Some(flow) = &change_password.flow {
+        Flow::remove(flow, &redis).await?;
+    }
+
+    if let Some(email) = user.email {
+        let changed = if update_password.is_some() {
+            "changed"
+        } else {
+            "removed"
+        };
+
+        send_email(
+            email,
+            &format!("Password {}", changed),
```

Comment:
More of a general thing, I don't think we need to send emails for things like password changes (removals, maybe). Is it just for extra safety? I don't really remember any other sites doing this

Reply:
I'm trying to think of a site where I don't get an email after changing my password (assuming they have my email in the first place)